client/can.go

```diff
@@ -0,0 +1,280 @@
+	Databases           []File `child:"file"`
+}
+
+// File represents a CAN database file in common formats such as KCD and DBC.
+type File struct {
+	Name string `point:"name"`
+	Data string `point:"data"`
+}
+
+// CanBusClient is a SIOT client used to communicate on a CAN bus
+type CanBusClient struct {
+	nc            *nats.Conn
+	config        CanBus
+	stop          chan struct{}
+	newPoints     chan NewPoints
+	newEdgePoints chan NewPoints
+	wrSeq         byte
+	lastSendStats time.Time
+	natsSub       string
+}
+
+// NewCanBusClient returns a new CanBusClient with a NATS connection and a config
+func NewCanBusClient(nc *nats.Conn, config CanBus) Client {
+	return &CanBusClient{
+		nc:            nc,
+		config:        config,
+		stop:          make(chan struct{}),
+		newPoints:     make(chan NewPoints),
+		newEdgePoints: make(chan NewPoints),
+		wrSeq:         0,
+		lastSendStats: time.Time{},
+		natsSub:       SubjectNodePoints(config.ID),
+	}
+}
+
+// Start runs the main logic for this client and blocks until stopped
+// There are several main aspects of the CAN bus client
+//
+//   - the listener function is a process that recieves CAN bus frames from the Linux
+//	   SocketCAN socket and sends the frames out on the canMsgRx channel
+//
+//   - when a frame is recieved on the canMsgRx channel in the main loop, it is decoded
+//	   and a point is sent out for each canparse.Signal in the frame. The key of each point
+//     contains the message name, signal name, and signal units
+//
+func (cb *CanBusClient) Start() error {
+	log.Println("CanBusClient: Starting CAN bus client: ", cb.config.Description)
+
+	var db *canparse.Database = &canparse.Database{}
+
+	readDb := func() {
+		cb.config.MsgsInDb = 0
+		cb.config.SignalsInDb = 0
+		db.Clean()
+		for _, dbFile := range cb.config.Databases {
+			err := db.ReadBytes([]byte(dbFile.Data), dbFile.Name)
+			if err != nil {
+				log.Println(errors.Wrap(err, "CanBusClient: Error parsing database file"))
+				return
+			} else {
+				/*
+					for _, b := range db.Busses {
+						cb.config.MsgsInDb += len(b.Messages)
+						for _, m := range b.Messages {
+							cb.config.SignalsInDb += len(m.Signals)
+							for _, s := range m.Signals {
+								log.Printf("CanBusClient: read msg %X sig %v: start=%v len=%v scale=%v offset=%v unit=%v",
+									m.Id, s.Name, s.Start, s.Length, s.Scale, s.Offset, s.Unit)
+							}
+						}
+					}
+				*/
+			}
+		}
+		points := data.Points{
+			data.Point{
+				Time:  time.Now(),
+				Type:  data.PointTypeMsgsInDb,
+				Value: float64(cb.config.MsgsInDb),
+			},
+			data.Point{
+				Time:  time.Now(),
+				Type:  data.PointTypeSignalsInDb,
+				Value: float64(cb.config.SignalsInDb),
+			},
+		}
+		err := SendPoints(cb.nc, cb.natsSub, points, false)
+		if err != nil {
+			log.Println(errors.Wrap(err, "CanBusClient: error sending points received from CAN bus: "))
+		}
+	}
+
+	readDb()
+
+	canMsgRx := make(chan can.Frame)
+
+	bringDownDev := func() {}
```

Comment:
could we fill in the bringDownDev function to close the socket? Otherwise, I think we will be leaking resources.

You could model after the closePort() in the serial client.